core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -66,140 +72,221 @@ class LogManager(val logDirs: Array[File],
   private val logs = new Pool[TopicPartition, Log]()
   private val logsToBeDeleted = new LinkedBlockingQueue[Log]()
 
-  createAndValidateLogDirs(logDirs)
-  private val dirLocks = lockLogDirs(logDirs)
-  private val recoveryPointCheckpoints = logDirs.map(dir => (dir, new OffsetCheckpointFile(new File(dir, RecoveryPointCheckpointFile)))).toMap
-  private val logStartOffsetCheckpoints = logDirs.map(dir => (dir, new OffsetCheckpointFile(new File(dir, LogStartOffsetCheckpointFile)))).toMap
+  val liveLogDirs: ArrayBuffer[File] = createAndValidateLogDirs(logDirs, initialOfflineDirs)
+  @volatile private var recoveryPointCheckpoints = liveLogDirs.map(dir => (dir, new OffsetCheckpointFile(new File(dir, RecoveryPointCheckpointFile)))).toMap
+  @volatile private var logStartOffsetCheckpoints = liveLogDirs.map(dir => (dir, new OffsetCheckpointFile(new File(dir, LogStartOffsetCheckpointFile)))).toMap
+
+  private def offlineLogDirs = logDirs.filterNot(liveLogDirs.toSet)
```

Comment:
Previously I think the chance of exception due to this race condition is so small (because disk failure should be rare) that we don't want to degrade performance by getting lock every time we read `liveLogDirs`. For example, metrics reporting will access `AbstractFetcherManager.fetcherThreadMap` and it seems fine so far.

But strictly speaking, you are right that this can cause race condition and it is better to prevent this completely from any unknown consequence. So I made the following changes to address the problem:

1) Use `_liveLogDirs: ArrayBlockingQueue[File]` to record offline log directories. We don't have to worry about blocking operation because we will only remove log directory from it. 

I think this may be a little better than using `logCreationOrDeletionLock synchronized {}` every time we access `_liveLogDirs` (e.g. `checkpointLogRecoveryOffsets()`, `checkpointLogStartOffsets` and metrics reporting) so that these access won't block `LogManager.createLog()` or `LogManager.asyncDelete()`

2) Add `def liveLogDirs: Array[File] = _liveLogDirs.asScala.toArray` so that we don't have to change test code to work with `ArrayBlockingQueue[File]`.